lib/rucio/transfertool/fts3.py

```diff
@@ -151,52 +158,123 @@ def submit(self, files, job_params, timeout=None):
+                    record_counter('transfertool.fts3.%s.submission.success' % self.__extract_host(self.external_host), len(files))
+                    labels = {'state': 'success', 'host': self.__extract_host(self.external_host)}
+                    SUBMISSION_COUNTER.labels(**labels).inc(len(files))
+                    transfer_id = str(post_result.json()['job_id'])
+                elif post_result and post_result.status_code == 409:
+                    record_counter('transfertool.fts3.%s.submission.failure' % self.__extract_host(self.external_host), len(files))
+                    labels = {'state': 'failure', 'host': self.__extract_host(self.external_host)}
+                    SUBMISSION_COUNTER.labels(**labels).inc(len(files))
+                    raise DuplicateFileTransferSubmission()
+                else:
+                    if expected_transfer_id:
+                        transfer_id = expected_transfer_id
+                        logging.warning("Failed to submit transfer to %s, will use expected transfer id %s, error: %s", self.external_host, transfer_id, post_result.text if post_result is not None else post_result)
+                    else:
+                        logging.warning("Failed to submit transfer to %s, error: %s", self.external_host, post_result.text if post_result is not None else post_result)
+                    record_counter('transfertool.fts3.%s.submission.failure' % self.__extract_host(self.external_host), len(files))
+                    labels = {'state': 'failure', 'host': self.__extract_host(self.external_host)}
+                    SUBMISSION_COUNTER.labels(**labels).inc(len(files))
+
+                if not transfer_id:
+                    raise TransferToolWrongAnswer('No transfer id returned by %s' % self.external_host)
+                return transfer_id
 
-        if post_result and post_result.status_code == 200:
-            record_counter('transfertool.fts3.%s.submission.success' % self.__extract_host(self.external_host), len(files))
-            labels = {'state': 'success', 'host': self.__extract_host(self.external_host)}
-            SUBMISSION_COUNTER.labels(**labels).inc(len(files))
-            transfer_id = str(post_result.json()['job_id'])
-        elif post_result and post_result.status_code == 409:
-            record_counter('transfertool.fts3.%s.submission.failure' % self.__extract_host(self.external_host), len(files))
-            labels = {'state': 'failure', 'host': self.__extract_host(self.external_host)}
-            SUBMISSION_COUNTER.labels(**labels).inc(len(files))
-            raise DuplicateFileTransferSubmission()
         else:
-            if expected_transfer_id:
-                transfer_id = expected_transfer_id
-                logging.warning("Failed to submit transfer to %s, will use expected transfer id %s, error: %s", self.external_host, transfer_id, post_result.text if post_result is not None else post_result)
+            # single VO mode
```

Comment:
This code block is identical to the one in the `if` part. Please move it to a function and call whenever needed.

Reply:
Thats what I thought when i copied and pasted, but there didnt seem to be similar methods, but will make this change

Reply:
Method abstracted